bonsai-core/src/main/scala/com/stripe/bonsai/IndexedBitSet.scala

```diff
@@ -0,0 +1,306 @@
+package com.stripe.bonsai
+
+import scala.annotation.switch
+import scala.collection.immutable.BitSet
+import scala.collection.mutable.{ Builder, ArrayBuilder }
+
+/**
+ * IndexedBitSet is an immutable set for storing non-negative integer
+ * values.
+ * 
+ * This data structure stores non-negative integers. It has O(1)
+ * lookup and rank operations. It also supports O(polylog n) select
+ * operations, which are not currently implemented.
+ *
+ * To support fast rank, this bitset requires ~1.37n bits, rather than
+ * the n bits that other bitsets (e.g. `scala.BitSet`) would use.
+ *
+ * Unlike some immutable structures, there is no form of structural
+ * sharing here. Thus, the only way to "modify" one of these
+ * structures is to build a new one that has the desired modifications
+ * (which will not be efficient if done frequently).
+ */
+final class IndexedBitSet(
+  val bits: Array[Int],
+  val length: Int,
+  val level2Start: Int,
+  val rawBitsStart: Int
+) {
+  import IndexedBitSet.rankWord
+
+  /**
+   * Test if `i` is contained in this bitset.
+   *
+   * The value `i` is expected to be non-negative.
+   */
+  def apply(i: Int): Boolean = {
```

Comment:
This data structure is based on a succinct data structure which, where the extra bits required for the index scales by log(n). I decided to fix n, since I felt that the complexity added by varying n wouldn't be worth some very minor gains. That said, the tl;dr is that 32-bits (and enforcing `Int` indices) gives us a good tradeoff for size and being able to pack the index in efficiently. I was going to refactor this a bit later on and may revisit allowing for larger bitsets (and smaller - but using fixed increments), but it's a bit of a can of worms.